lxd/acme.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/lxc/lxd/lxd/db/operationtype"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/lxd/task"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+var apiACME = []APIEndpoint{
+	acmeChallengeCmd,
+}
+
+var acmeChallengeCmd = APIEndpoint{
+	Path: ".well-known/acme-challenge/{token}",
+
+	Get: APIEndpointAction{Handler: acmeProvideChallenge, AllowUntrusted: true},
+}
+
+func acmeProvideChallenge(d *Daemon, r *http.Request) response.Response {
+	token, err := url.PathUnescape(mux.Vars(r)["token"])
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// If we're clustered, forwared the request to the leader if necessary.
+	// That is because only the leader knows the token, and any other node will return 404.
+	clustered, err := cluster.Enabled(d.db.Node)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if clustered {
+		leader, err := d.gateway.LeaderAddress()
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// This gives me the correct value
+		clusterAddress := d.State().LocalConfig.ClusterAddress()
+
+		if clusterAddress != "" && clusterAddress != leader {
+			// Forward the request to the leader
+			client, err := cluster.Connect(leader, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			return response.ForwardedResponse(client, r)
+		}
+	}
+
+	if d.http01Provider == nil || d.http01Provider.Token() != token {
+		return response.NotFound(nil)
+	}
+
+	return response.ManualResponse(func(w http.ResponseWriter) error {
+		w.Header().Set("Content-Type", "text/plain")
+
+		_, err := w.Write([]byte(d.http01Provider.KeyAuth()))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func autoRenewServerCert(ctx context.Context, d *Daemon) error {
+	domain, email, caURL := d.globalConfig.ACME()
+
+	if domain == "" || email == "" {
+		return nil
+	}
+
+	clustered, err := cluster.Enabled(d.db.Node)
+	if err != nil {
+		return err
+	}
+
+	var clusterAddress string
+
+	// If we are clustered, let the leader handle the certificate renewal.
+	if clustered {
+		leader, err := d.gateway.LeaderAddress()
+		if err != nil {
+			return err
+		}
+
+		// Figure out our own cluster address.
+		clusterAddress := d.State().LocalConfig.ClusterAddress()
+
+		if clusterAddress != leader {
+			return nil
+		}
+	}
+
+	opRun := func(op *operations.Operation) error {
+		cert, err := acme.UpdateACMECertificate(d.State(), d.http01Provider, clustered, domain, email, caURL)
+		if err != nil {
+			return err
+		}
+
+		// If cert is nil, there's no need to update it as it's still valid for more than 30 days.
+		if cert == nil {
+			return nil
+		}
+
+		req := api.ClusterCertificatePut{
+			ClusterCertificate:    string(cert.Certificate),
+			ClusterCertificateKey: string(cert.PrivateKey),
+		}
+
+		if clustered {
+			servers, err := d.gateway.NodeStore().Get(context.Background())
```

Comment:
@monstermunchkin this doesn't look right. This only returns database nodes, what about LXD cluster members that are not database nodes?

Normally we use  `tx.GetNodes(ctx)` for this sort of operation.